lib/api/apiUtils/bucket/bucketWebsite.js

```diff
@@ -0,0 +1,275 @@
+import { parseString } from 'xml2js';
+
+import { errors } from 'arsenal';
+import {
+    RoutingRule,
+    WebsiteConfiguration,
+} from '../../../metadata/WebsiteConfiguration';
+/*
+   Format of xml request:
+
+   <WebsiteConfiguration xmlns='http://s3.amazonaws.com/doc/2006-03-01/'>
+     <IndexDocument>
+       <Suffix>index.html</Suffix>
+     </IndexDocument>
+     <ErrorDocument>
+       <Key>Error.html</Key>
+     </ErrorDocument>
+
+     <RoutingRules>
+       <RoutingRule>
+       <Condition>
+         <KeyPrefixEquals>docs/</KeyPrefixEquals>
+       </Condition>
+       <Redirect>
+         <ReplaceKeyPrefixWith>documents/</ReplaceKeyPrefixWith>
+       </Redirect>
+       </RoutingRule>
+       ...
+    </RoutingRules>
+   </WebsiteConfiguration>
+   */
+
+
+/** Check if parsed xml element contains a specified child element
+* @param {obj} obj - represents xml element to check for child element
+* @param {string} requiredElem - name of child element
+* @param {boolean} isList - indicates if obj is list of children elements
+* @return {boolean} true / false - if parsed xml element contains child
+*/
+function _xmlContainsElem(obj, requiredElem, isList) {
+    if (!Array.isArray(obj)
+    || obj.length !== 1
+    || !Array.isArray(obj[0][requiredElem])) {
+        return false;
+    }
+    if (!isList) {
+        if (obj[0][requiredElem].length !== 1) {
+            return false;
+        }
+    } else if (obj[0][requiredElem].length === 0) {
+        return false;
+    }
+
+    return true;
+}
+
+/** Validate XML, returning an error if any part is not valid
+* @param {object} jsonResult - website configuration xml parsed into JSON
+* @return {Error} on error, {WebsiteConfiguration} on success
+*/
+function _validateWebsiteConfigXml(jsonResult) {
```

Comment:
I can split them up into smaller functions correlating to xml elements. I agree it's an `if` forest. :P Can you describe what you're imagining for a state-machine? Could that be managed by an object mapping to functions? How to get around using all the `if` statements?